aten/src/ATen/native/Gesv.cpp

```diff
@@ -0,0 +1,123 @@
+#include "ATen/ATen.h"
+#include "ATen/CPUApplyUtils.h"
+#include "ATen/Dispatch.h"
+#include "ATen/ExpandUtils.h"
+#include "ATen/NativeFunctions.h"
+
+#include "ATen/native/LinearAlgebraUtils.h"
+#include "ATen/native/Gesv.h"
+
+#include "TH.h"  // for USE_LAPACK
+
+#include <vector>
+
+#ifdef USE_LAPACK
+extern "C" void dgesv_(
+    int* n, int* nrhs, double* a, int* lda,
+    int *ipiv, double* b, int* ldb, int* info);
+extern "C" void sgesv_(
+    int* n, int* nrhs, float* a, int* lda,
+    int* ipiv, float* b, int* ldb, int* info);
+#endif
+
+namespace at { namespace native {
+
+template<class real>
+void lapackGesv(
+    int n, int nrhs, real* a, int lda, int* ipiv,
+    real* b, int ldb, int* info) {
+  AT_ERROR("gesv only takes float or double Tensors");
+}
+
+#ifdef USE_LAPACK
+template<> void lapackGesv<float>(
+    int n, int nrhs, float* a, int lda, int* ipiv,
+    float* b, int ldb, int* info) {
+  sgesv_(&n, &nrhs, a, &lda, ipiv, b, &ldb, info);
+}
+
+template<> void lapackGesv<double>(
+    int n, int nrhs, double* a, int lda, int* ipiv,
+    double* b, int ldb, int* info) {
+  dgesv_(&n, &nrhs, a, &lda, ipiv, b, &ldb, info);
+}
+#endif
+
+template <typename real>
+static void applyGesv(Tensor& b, Tensor& A, std::vector<int64_t> infos) {
+#ifndef USE_LAPACK
+  AT_ERROR("gesv: LAPACK library not found in compilation");
+#endif
+  real* A_data = (real*)A.data_ptr();
+  real* b_data = (real*)b.data_ptr();
+  auto A_mat_stride = matrixStride(A);
+  auto b_mat_stride = matrixStride(b);
+
+  auto batch_size = batchCount(A);
+  auto n = A.size(-2);
+  auto nrhs = b.size(-1);
+
+  auto ipiv = b.type().toScalarType(kLong).tensor(n);
+
+  for (int64_t i = 0; i < batch_size; i++) {
+    int info;
+    real* A_working_ptr = &A_data[i * A_mat_stride];
+    real* b_working_ptr = &b_data[i * b_mat_stride];
+    lapackGesv<real>(n, nrhs, A_working_ptr, n, (int*)ipiv.data_ptr(),
```

Comment:
I'm just curious. If we are keeping the TH code, why can't we just call TH `_gesv_single` here? Is this code doing something different with the TH gesv, or is it here so that migrating the entire thing to ATen will be easier?

Reply:
Using gesv_single would make the error messages confusing to the user. What I’d really like is to use the THLapack_(gesv) function, but that isn’t exposed to ATen, so I made a port of it

Reply:
Thanks for explaining. Let's add some comments here then, and also a TODO on moving gesv to ATen completely.